### PR TITLE
TST: Add tests for 4bit LoftQ

### PR DESCRIPTION
**Pull request by BenjaminBossan**
Add GPU tests for LoftQ with 4bit quantization.

## Notes

Tests for 8bit quantization are already there but not run at the moment, see this comment:

https://github.com/huggingface/peft/pull/1150#issuecomment-1838891499

In my testing, 8bit passes when using `NFQuantizer`, so if the original author is fine with using that, I can make the adjustment.

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
@yxli2123 If you could also do a quick review, that would be great.

I ran the GPU tests locally and they passed, so :crossed_fingers: that they also pass on the nightly CI (PR CI does not run GPU tests).

File: tests/test_gpu_examples.py [hunk 1]
```diff
@@ -933,3 +935,138 @@ def test_causal_lm_training_mutli_gpu(self):
...
             # assert loss is not None
             self.assertIsNotNone(trainer.state.log_history[-1]["train_loss"])
+
+
+@require_torch_gpu
+class LoftQTests(unittest.TestCase):
+    r"""
+    Tests for LoftQ
+    """
+
```
Comment (younesbelkada) [inline tests/test_gpu_examples.py:945]:
```suggestion
    error_factor = 3
```

File: tests/test_gpu_examples.py [hunk 2]
```diff
@@ -933,3 +935,138 @@ def test_causal_lm_training_mutli_gpu(self):
...
+        # first, sanity check that all errors are > 0.0
+        self.assertTrue(mae_quantized > 0.0)
+        self.assertTrue(mse_quantized > 0.0)
+        self.assertTrue(mae_loftq > 0.0)
+        self.assertTrue(mse_loftq > 0.0)
+
+        # next, check that LoftQ quantization errors are smaller than LoRA errors by a certain margin
+        factor = 3
+        self.assertTrue(mae_loftq < mae_quantized / factor)
+        self.assertTrue(mse_loftq < mse_quantized / factor)
```
Comment (younesbelkada) [inline tests/test_gpu_examples.py:1032]:
```suggestion
        self.assertTrue(mae_loftq < mae_quantized / self.error_factor)
        self.assertTrue(mse_loftq < mse_quantized / self.error_factor)
```

File: tests/test_gpu_examples.py [hunk 3]
```diff
@@ -933,3 +935,138 @@ def test_causal_lm_training_mutli_gpu(self):
...
```
Comment (younesbelkada) [inline tests/test_gpu_examples.py:1052]:
```suggestion
        self.assertTrue(mae_loftq < mae_quantized / self.error_factor)
        self.assertTrue(mse_loftq < mse_quantized / self.error_factor)
```

File: tests/test_gpu_examples.py [hunk 4]
```diff
@@ -933,3 +935,138 @@ def test_causal_lm_training_mutli_gpu(self):
...
```
Comment (younesbelkada) [inline tests/test_gpu_examples.py:1072]:
```suggestion
        self.assertTrue(mae_loftq < mae_quantized / self.error_factor)
        self.assertTrue(mse_loftq < mse_quantized / self.error_factor)
```

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for adding the test ! I left one suggestion so that we can easily play with the error tolerance in the future in case the test fail on the runner - what do you think?

**Pull request merged**